### DISCUSSION TITLE
Leftover database tables from uninstalled plugins

### DISCUSSION THREAD

**Discussion by alfredbuehler** [aljawaid/ContentCleaner#2] [Ideas]
I often wanted a tool to remove such waste. OTOH, sometimes it's better, not to remove, in case you reinstall.

**Comment by aljawaid** [comment]
I agree, you originally highlighted the problem a few months ago and now I understand it better.  I'm trying to think of a few ways around it.  I am creating my first plugin with a database, after that, I will look into this for sure as I will want to delete my own tables too after uninstalling.

**Reply by aljawaid** [reply]
#1 Concept

**Reply by aljawaid** [reply]
Do you know how to create a variable for the table name?

I have a function which deletes a specified name, but I cant get the variable to work.  Its not practical I think to create a delete function and form for every plugin which has a database table.  Any tips? @alfredbuehler  @creecros

I created the model, to be called by the controlller. The `remove` form should call the controller (which pulls the sql command through the model) to delete the table specified.  That's my thinking, in case I am wrong.

I just need somebody to look at the code and tell me how I am wrong.

**Reply by creecros** [reply]
I might have done about 5 mins of thought experiment on this, so maybe there is a better way, and what I have is lengthy.

1. using kb schema, compile a list of all the tables that are standard.
2. using schemas of current plugins installed, do the same.
3. combine the above into array of names.
4. compile list of all tables in database as array of names.
5. compare these 2 arrays, and determine which tables are not needed. compile into new array of names.
6. delete those tables.

**Reply by creecros** [reply]
that gets even more complicated as i think about it, because you will also need the corresponding column names....and account for all the alterations to the names throughout the schema....

**Reply by aljawaid** [reply]
woooah... I was simply trying to delete tables by checking which plugin has a schema.  If i compare through your way, I think it would be too automated and potentially deleting data which might not want to be deleted.

Will my code not work?  I got as far as deleting a 'test' table.  I just need to put a variable in it so the table name can be changed through each button displayed.

**Reply by creecros** [reply]
and you are saying what?
https://github.com/aljawaid/ContentCleaner/blob/802032a4f6b4041cda8228148e8f96b351568527/Model/ApplicationCleaningModel.php#L39

that does not work? 

back to my previous, if the code is correct, there is 0 potential in your concern. There would be way more concern that humans make more mistakes and delete the wrong table.

**Reply by aljawaid** [reply]
> and you are saying what?
> https://github.com/aljawaid/ContentCleaner/blob/802032a4f6b4041cda8228148e8f96b351568527/Model/ApplicationCleaningModel.php#L39
> 
> that does not work? 
> 
> back to my previous, if the code is correct, there is 0 potential in your concern. There would be way more concern that humans make more mistakes and delete the wrong table.

That's what I don't know though, if the code is correct... If its correct in the model, is it also correct in the controller and the button?

**Reply by creecros** [reply]
im not following. test it.

**Reply by aljawaid** [reply]
I have tested, I get the modal, but the confirm buttons just keep spinning and as a result (in phpmyadmin) the `test` table is not deleted.  There is a problem with the variable somewhere.

```
<a id="" href="<?= $this->url->href('CleaningController', 'confirm', array('table' => 'test', 'plugin' => 'ContentCleaner'), false, '', false) ?>" class="btn  js-modal-confirm" title="<?=t('Delete') ?>">
```
**The above button works, it pulls up the modal.**

Button in `remove.php` modal:
```php
<?= $this->modal->confirmButtons(
        'CleaningController',
        'removeTable',
        array($table => 'table', 'plugin' => 'ContentCleaner'),
        t('Delete')
    ) ?>
```
**The above button in the modal just keeps spinning**

So i'm guessing, either the model is wrong:
```php
public function delete($table)
    {
        // delete all

        return $this->db->execute('DROP TABLE IF EXISTS `'.$table.'`; SHOW WARNINGS');
    }
```

or the controller:
```php
public function removeTable()
    {
        $this->checkCSRFParam();

        if ($this->applicationCleaningModel->delete()) {
            $this->flash->success(t('Cleaning complete - Database table deleted successfully'));
        } else {
            $this->flash->failure(t('Cleaning failed'));
        }

        $this->response->redirect($this->helper->url->to('ContentCleanerController', 'show', array('plugin' => 'ContentCleaner')), true);
    }
```
I feel like I am missing passing the variable somewhere.

**Reply by creecros** [reply]
you didnt pass the variable in your remove.php

**Reply by creecros** [reply]
i take that back, you did. but you are not picking it up correctly. need to getParam() or getStringParam(), or getIntegerParam()

here is an example:
https://github.com/creecros/Mailmagik/blob/29872cd0fce54757fcc248ac2b15ec9c15c34251/Controller/EmailViewController.php#L149

**Reply by aljawaid** [reply]
`$table =  $this->request->getStringParam('table');`  in the model? or in the controller? or in the remove.php?

**Reply by aljawaid** [reply]
oh wait, you put yours in the controller, let me try that

**Reply by creecros** [reply]
you already passed it in your modal. thats good. :heavy_check_mark: 

your current controller is not picking it up. thats bad. thats where you need to pick it up. :x:
i also do not see where you pass it to the model. thats also bad. it wont know what to delete. :x:

**Reply by creecros** [reply]
Ill just update your code...

> ```
> <a id="" href="<?= $this->url->href('CleaningController', 'confirm', array('table' => 'test', 'plugin' => 'ContentCleaner'), false, '', false) ?>" class="btn  js-modal-confirm" title="<?=t('Delete') ?>">
> ```
> 
> **The above button works, it pulls up the modal.**
> 
> Button in `remove.php` modal:
> 
> ```
> <?= $this->modal->confirmButtons(
>         'CleaningController',
>         'removeTable',
>         array($table => 'table', 'plugin' => 'ContentCleaner'),
>         t('Delete')
>     ) ?>
> ```
> 
> **The above button in the modal just keeps spinning**
> 
> So i'm guessing, either the model is wrong:
> 
> ```
> public function delete($table)
>     {
>         // delete all
> 
>         return $this->db->execute('DROP TABLE IF EXISTS `'.$table.'`; SHOW WARNINGS');
>     }
> ```
> 
> or the controller:
> 
> ```
> public function removeTable()
>     {
>         $table =  $this->request->getStringParam('table')
>         $this->checkCSRFParam();
> 
>         if ($this->applicationCleaningModel->delete($table)) {
>             $this->flash->success(t('Cleaning complete - Database table deleted successfully'));
>         } else {
>             $this->flash->failure(t('Cleaning failed'));
>         }
> 
>         $this->response->redirect($this->helper->url->to('ContentCleanerController', 'show', array('plugin' => 'ContentCleaner')), true);
>     }
> ```
> 
> I feel like I am missing passing the variable somewhere.

**Reply by creecros** [reply]
oh i see another mistake

```
<?= $this->modal->confirmButtons(
        'CleaningController',
        'removeTable',
        array($table => 'table', 'plugin' => 'ContentCleaner'),
        t('Delete')
    ) ?>
```

`$table` is flipped with `'table'`  should be `'table' => $table`

**Reply by aljawaid** [reply]
after your code update, thanks for that... but its still spinning...

`removeTable()` is coded how you mentioned above... but its still spinning in the modal without a result or error.

is this controlller function correct?
```
public function confirm()
    {
        $this->response->html($this->template->render('contentCleaner:config/remove', array(
            'table' => $this->request->getStringParam('table'),     ////this gives no error, no result
//or should it be             'table' => $table,    // this gives undefined variable but still no result
        )));
    }
```

**Reply by creecros** [reply]
your gonna make me download this and figure it out arent you. give me a sec.

**Reply by aljawaid** [reply]
> oh i see another mistake
> 
> ```
> <?= $this->modal->confirmButtons(
>         'CleaningController',
>         'removeTable',
>         array($table => 'table', 'plugin' => 'ContentCleaner'),
>         t('Delete')
>     ) ?>
> ```
> 
> `$table` is flipped with `'table'` should be `'table' => $table`

that worked! and also the below:
```
public function confirm()
    {
        $this->response->html($this->template->render('contentCleaner:config/remove', array(
            'table' => $this->request->getStringParam('table'),
            //'table' => $table,
        )));
    }
```


THANKS SO MUCH

**Reply by aljawaid** [reply]
> your gonna make me download this and figure it out arent you. give me a sec.

no need, you fixed it :) tabel got deleted.... now I can create different buttons for different plugins and different sql queries for core tables to cleanup data.... @alfredbuehler looks like thanks to creecros, thsi plugin will see a release :)

**Reply by creecros** [reply]
aborting download :stop_sign:

**Comment by creecros** [comment]
Plausible start, had ChatGPT help me on this one. Haven't actually tested it. (ChatGPT could be full of shit though)

```
function getTableColumns($schema_file_path) {
  // Load the schema file
  $schema = file_get_contents($schema_file_path);

  // Split the schema into individual statements
  $statements = preg_split('/\s*;\s*/', $schema, -1, PREG_SPLIT_NO_EMPTY);

  // Initialize an array to store the table names and column names
  $table_columns = array();

  // Loop through each statement
  foreach ($statements as $statement) {
    // Check if the statement is a CREATE TABLE statement
    if (preg_match('/^CREATE\s+TABLE\s+`?(\w+)`?/i', $statement, $matches)) {
      // Extract the table name
      $table_name = $matches[1];

      // Extract the column names
      preg_match_all('/`?(\w+)`?\s+\w+\s*(?:\(\s*\d+\s*\))?\s*(?:NOT\s+NULL)?\s*(?:AUTO_INCREMENT)?(?:\s*DEFAULT\s*\'?\w+\'?)?/i', $statement, $matches);
      $column_names = $matches[1];

      // Add the table name and column names to the result array
      $table_columns[$table_name] = $column_names;
    }

    // Check if the statement is an ALTER TABLE statement for dropping a column
    if (preg_match('/^ALTER\s+TABLE\s+`?(\w+)`?\s+DROP\s+COLUMN\s+`?(\w+)`?/i', $statement, $matches)) {
      // Remove the column name from the corresponding table's column list
      $table_name = $matches[1];
      $column_name = $matches[2];
      $key = array_search($column_name, $table_columns[$table_name]);
      if ($key !== false) {
        unset($table_columns[$table_name][$key]);
      }
    }

    // Check if the statement is an ALTER TABLE statement for renaming a column
    if (preg_match('/^ALTER\s+TABLE\s+`?(\w+)`?\s+CHANGE\s+`?(\w+)`?\s+`?(\w+)`?\s+\w+\s*(?:\(\s*\d+\s*\))?\s*(?:NOT\s+NULL)?\s*(?:AUTO_INCREMENT)?(?:\s*DEFAULT\s*\'?\w+\'?)?/i', $statement, $matches)) {
      // Rename the column in the corresponding table's column list
      $table_name = $matches[1];
      $old_column_name = $matches[2];
      $new_column_name = $matches[3];
      $key = array_search($old_column_name, $table_columns[$table_name]);
      if ($key !== false) {
        $table_columns[$table_name][$key] = $new_column_name;
      }
    }

    // Check if the statement is a DROP TABLE statement
    if (preg_match('/^DROP\s+TABLE\s+`?(\w+)`?/i', $statement, $matches)) {
      // Remove the table and its column names from the result array
      $table_name = $matches[1];
      unset($table_columns[$table_name]);
    }
  }

  return $table_columns;
}
```

**Reply by aljawaid** [reply]
besides the comments, the rest is a bit complicated lol.... but too complicated for me to do if I dont understand it at all.

**Reply by creecros** [reply]
i dont understand the point of the plugin if the user has to decide what to delete. thats scary as hell.

**Reply by creecros** [reply]
chatgpt lied anyway. shit dont work.
